### PR TITLE
 bpo-31374: Create expat_config.h with autoconf

### PR DESCRIPTION
**Pull request by tiran**
include expat_config.h first to avoid warning:

  "_POSIX_C_SOURCE" redefined

Signed-off-by: Christian Heimes <christian@python.org>

https://bugs.python.org/issue31374

### CODE REVIEW

**Comment by vstinner** [comment]
Would it be possible to modify expat_config.h instead of modifying code copied from libexpat?

**Comment by vstinner** [comment]
The AppVeyor failure is unrelated to your change, it's a regression that I introduced when I made threading_cleanup() stricter: https://bugs.python.org/issue31234#msg302163

**Comment by tiran** [comment]
We could fix it in ``configure.ac`` by replacing ``AC_DEFINE`` for the macro to:
```
AH_VERBATIM([_POSIX_C_SOURCE],
[/* Define to activate features from IEEE Stds 1003.1-2008 */
#ifdef _POSIX_C_SOURCE
#undef _POSIX_C_SOURCE
#endif
#define _POSIX_C_SOURCE 200809L
])
```

**Comment by vstinner** [comment]
Hum, I suggest to not define _POSIX_C_SOURCE if it's already set, rather than replacing the value:

```
#ifdef _POSIX_C_SOURCE
#  define _POSIX_C_SOURCE 200809L
#endif
```

Do we really need pyconfig.h to configure expat?

**Comment by tiran** [comment]
We use pyconfig.h to get endianess.

**Comment by vstinner** [comment]
I tried to generate expat_config.h from a expat_config.h.in template which doesn't include pyconfig.h. 

Problem: Expat requires the endian, but also the HAVE_MEMMOVE define. I don't know autotools good enough to propose a working change.

I tried the following code in configure.ac:

```
# Define a BYTEORDER variable for Moules/expat/expat_config.h.in
AC_SUBST(BYTEORDER)
AC_C_BIGENDIAN(
  BYTEORDER=4321,
  BYTEORDER=1234,
  AC_MSG_ERROR(unknown endianess),
  AC_MSG_ERROR(universial endianess not supported)
)
```

but then noticed that autoheader genereates more complex code in pyconfig.h for the endian:

```
/* Define WORDS_BIGENDIAN to 1 if your processor stores words with the most
   significant byte first (like Motorola and SPARC, unlike Intel). */
#if defined AC_APPLE_UNIVERSAL_BUILD
# if defined __BIG_ENDIAN__
#  define WORDS_BIGENDIAN 1
# endif
#else
# ifndef WORDS_BIGENDIAN
#  undef WORDS_BIGENDIAN
# endif
#endif
```

Oh wow, I didn't know that getting the endian would be so complex.

I don't know how to pass HAVE_MEMMOVE from configure.ac to my Modules/expat/expat_config.h.in.

Well, as @tiran told me on IRC, my solution is maybe too complex just to fix a single compiler warning :-)

--

But I dislike modifying pyconfig.h because portability is also complex, and I don't know what are the consequences. I proposed a different fix which leaves pyconfig.h unchanged: PR #3601. It also documents what are get from pyconfig.h.

@tiran: What do you think?

File: configure.ac [hunk 1]
```diff
@@ -4585,6 +4582,26 @@ AC_MSG_RESULT($HAVE_USABLE_WCHAR_T)
...
+        AC_MSG_ERROR([unknown endianness $ac_cv_c_bigendian])
+        ;;
+esac
+AC_SUBST([BYTEORDER])
+
+AC_CHECK_FUNCS(memmove,
+  [DEFINE_HAVE_MEMMOVE='#define HAVE_MEMMOVE 1'],
+  [DEFINE_HAVE_MEMMOVE='#undef HAVE_MEMMOVE'],
+)
+AC_SUBST([DEFINE_HAVE_MEMMOVE])
```
Comment (vstinner) [inline configure.ac:4557]: You may define AC_CONFIG_FILES([Modules/expat/expat_config.h]) here, to group things specific to Expat at the same place. It seems like AC_CONFIG_FILES() can be called multiple times.

File: Misc/NEWS.d/next/Build/2017-09-14-11-34-13.bpo-31374.VFnngi.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+pyconfig.h no longer defines _POSIX_C_SOURCE when it is already defined. This
+fixes a build warning in Python's copy of expat library.
```
Comment (vstinner) [inline Misc/NEWS.d/next/Build/2017-09-14-11-34-13.bpo-31374.VFnngi.rst]: This NEWS entry now seems to be outdated.

**Review by vstinner — CHANGES_REQUESTED** [review]
You forgot the newly added .in file, no?

File: configure.ac (hunk 1 above)
Comment (tiran) [reply configure.ac:4557]: I prefer to keep all output files in one place

**Comment by tiran** [comment]
I didn't expect the **French** Inquisition!

And I didn't expect the Spanish Inquisition!

**Comment by vstinner** [comment]
I tested this PR when HAVE_MEMMOVE is false (undefined): expat compilation fails because HAVE_BCOPY is unset, whereas my Fedora 29 supports bcopy() :-/

libexpat uses more "HAVE_xxx" macros:

```
HAVE_ARC4RANDOM
HAVE_ARC4RANDOM_BUF
HAVE_BCOPY
HAVE_EXPAT_CONFIG_H
HAVE_GETRANDOM
HAVE_LIBBSD
HAVE_SYSCALL_GETRANDOM
```

expat_config.h seems like the best approach, but I'm not longer sure that it's worth it. This approach requires a lot of work, whereas most platforms use system libexpat, not our vendored libexpat copy.

**Comment by vstinner** [comment]
This approach seems too complex :-( Since we are trying to fix a minor warning, I wrote a simpler patch which modifies the vendored copy of libexpat: PR #11064.

**Comment by vstinner** [comment]
I merged a simpler approach: PR #11064.

**Pull request closed without merging**